### PR TITLE
UTY-835; Invalidate deinjected requirables

### PR DESCRIPTION
**Pull request by JonasImprobable**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
- Requirables can now be disposed. Once disposed you cannot access them anymore. (Read component state, register/deregister callbacks).
- Requirables are disposed upon losing a component/authority and before the gameobject is removed.
- Removed MonoBehaviourActivationManagerInitializationSystem and merged its logic into the EntityGameObjectLinker.
- You can now Unlink gaemobjects.
#### Tests
Only manual tests for now for the sake of shipping this in time. Proper tests shall be done after the alpha release
#### Documentation
To be done in a follow-up PR.
#### Primary reviewers
@MPeti01 @firtina-improbable

### CODE REVIEW

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectDispatcherSystem.cs [hunk 1]
```diff
@@ -177,5 +166,18 @@ public void CreateActivationManagerAndReaderWriterStore(Entity entity)
             var manager = new MonoBehaviourActivationManager(gameObject, injector, store, logger);
             entityToActivationManager.Add(entity, manager);
         }
+
+        internal void RemoveActivationManagerAndReaderWriterStore(Entity entity)
+        {
+            if (!entityToActivationManager.ContainsKey(entity))
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectDispatcherSystem.cs]: Use `TryGet` to avoid a double lookup here and below.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/MonoBehaviourActivationManager.cs [hunk 2]
```diff
@@ -105,25 +105,48 @@ public void EnableSpatialOSBehaviours()
             foreach (var behaviour in behavioursToEnable)
             {
                 behaviour.enabled = true;
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/MonoBehaviourActivationManager.cs]: Can exceptions propagate out of the `OnEnable` method of the MonoBehaviour?

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/MonoBehaviourActivationManager.cs [hunk 3]
```diff
@@ -105,25 +105,48 @@ public void EnableSpatialOSBehaviours()
...
+            foreach (var behaviour in behaviours)
             {
                 injector.DeInjectAllRequiredFields(behaviour);
                 store.RemoveInjectablesForBehaviour(behaviour);
             }
 
-            behavioursToDisable.Clear();
+            // Make a copy because the behaviours argument could potentially be the enabledBehaviours set itself.
+            var removedBehaviours = new List<MonoBehaviour>(behaviours);
+            foreach (var removedBehaviour in removedBehaviours)
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/MonoBehaviourActivationManager.cs]: Should these be removed in reverse order?

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs [hunk 4]
```diff
@@ -68,6 +68,26 @@ private IInjectable Inject(MonoBehaviour behaviour, InjectableId injectableId, E
...
+            EnsureLoaded(behaviourType);
+            foreach (var fieldsToComponents in fieldInfoCache[behaviourType])
+            {
+                var fields = fieldsToComponents.Value;
+                foreach (var field in fields)
+                {
+                    var oldField = field.GetValue(behaviour);
+                    if (!(oldField is RequirableBase))
+                    {
+                        Debug.LogError("Something is wrong " + behaviour);
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs]: Gonna need to be more specific that that :)

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs [hunk 5]
```diff
@@ -68,6 +68,26 @@ private IInjectable Inject(MonoBehaviour behaviour, InjectableId injectableId, E
...
+        {
+            var behaviourType = behaviour.GetType();
+            EnsureLoaded(behaviourType);
+            foreach (var fieldsToComponents in fieldInfoCache[behaviourType])
+            {
+                var fields = fieldsToComponents.Value;
+                foreach (var field in fields)
+                {
+                    var oldField = field.GetValue(behaviour);
+                    if (!(oldField is RequirableBase))
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs]: Case it once and save the result so you don't have to cast again below.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReadersWriters/RequirableBase.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        public RequirableBase(ILogDispatcher logDispatcher)
+        {
+            this.logDispatcher = logDispatcher;
+        }
+
+        protected bool LogErrorIfDisposed()
+        {
+            if (isDisposed)
+            {
+                logDispatcher.HandleLog(LogType.Error, new LogEvent(AccessedDisposedRequirable)
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReadersWriters/RequirableBase.cs]: I can imagine this becoming very spammy if this code path fires. Any ideas for how to mitigate this?

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectDispatcherSystem.cs (hunk 1 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectDispatcherSystem.cs]: Fixed. :)

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/MonoBehaviourActivationManager.cs (hunk 2 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/MonoBehaviourActivationManager.cs]: Thanks for catching! Fixed.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs (hunk 5 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs]: Fixed.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReadersWriters/RequirableBase.cs (hunk 6 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReadersWriters/RequirableBase.cs]: Yeah, I can see how this could get spammy. We could save for which [Require] field or MonoBehaviour type we've already logged an error and only log an error once per field/Behaviour but that would mean that we potentially accumulate a bit cache over time. Do you have a suggestion for how we could handle this more elegantly?

File: workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/EntityGameObjectLinkerTests.cs [hunk 7]
```diff
@@ -23,6 +24,7 @@ public void Setup()
         {
             world = new World("TestWorld");
             entityManager = world.GetOrCreateManager<EntityManager>();
+            world.CreateManager<WorkerSystem>(null, null, "TestWorker", Vector3.zero);
```
Comment (jessicafalk) [inline workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/EntityGameObjectLinkerTests.cs]: do we need that line?

Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.core/Tests/Editmode/GameObjectRepresentation/EntityGameObjectLinkerTests.cs]: Unfortunately yes. Upon creating the `EntityGameObjectLinker`, that object will getorcreate a `GameObjectDispatcherSystem` which will attempt to call `World.GetExistingManager<WorkerSystem>()`

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs [hunk 8]
```diff
@@ -43,6 +45,30 @@ public EntityGameObjectLinker(World world, ILogDispatcher logDispatcher)
...
             spatialOSComponent.World = world;
             spatialOSComponent.LogDispatcher = logDispatcher;
+
+            var gameObjectReference = new GameObjectReference { GameObject = gameObject };
+            viewCommandBuffer.AddComponent(entity, gameObjectReference);
+
+            var gameObjectReferenceHandleComponent = new GameObjectReferenceHandle();
+            entityCommandBuffer.AddComponent(entity, gameObjectReferenceHandleComponent);
+
+            gameObjectDispatcherSystem.CreateActivationManagerAndReaderWriterStore(entity, gameObject);
```
Comment (jessicafalk) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs]: I don't find it very intuitive that we call this from a system to call into a different system.  Systems should be modular by itself and there should be no need for a system to reference adifferent one (even indirect).
It would be better to factor this out into two different systems. It looks like we try to much in one system

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs [hunk 9]
```diff
@@ -11,15 +11,17 @@ public class EntityGameObjectLinker
...
         public EntityGameObjectLinker(World world, ILogDispatcher logDispatcher)
         {
             this.world = world;
             this.logDispatcher = logDispatcher;
+            gameObjectDispatcherSystem = world.GetOrCreateManager<GameObjectDispatcherSystem>();
         }
 
         public void LinkGameObjectToEntity(GameObject gameObject, Entity entity, EntityId spatialEntityId,
-            ViewCommandBuffer viewCommandBuffer)
+            EntityCommandBuffer entityCommandBuffer, ViewCommandBuffer viewCommandBuffer)
```
Comment (jessicafalk) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs]: why do we need to add two buffers?

Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs]: One is for adding Componentdatas and one is for adding non-blittable components.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs [hunk 10]
```diff
@@ -27,6 +27,8 @@ internal class RequiredFieldInjector
             = "[Require] attribute found on member that is not Injectable. This member will be ignored.";
         private const string MalformedInjectable
             = "Injectable found without required attributes, this is invalid.";
+        private const string RequirableFieldDoesNotInheritRequirableBase
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs]: Wondering if this can be tested in an editor test

Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs]: @jessicafalk can it be done similarly to how you tested the systems annotations?

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReadersWriters/ReaderWriterBase.cs [hunk 11]
```diff
@@ -9,16 +9,15 @@
 namespace Improbable.Gdk.Core.GameObjectRepresentation
 {
     internal abstract class ReaderWriterBase<TSpatialComponentData, TComponentUpdate>
-        : IWriter<TSpatialComponentData, TComponentUpdate>
+        : RequirableBase
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReadersWriters/ReaderWriterBase.cs]: Won't removing the `IWriter` inheritance cause problems?

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs (hunk 10 above)
Comment (jessicafalk) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs]: that should be doable with a bit of reflection magic

File: workers/unity/Assets/Generated/Source/improbable/gdk/tests/ExhaustiveBlittableSingularReaderWriter.cs [hunk 12]
```diff
@@ -64,120 +64,360 @@ public ReaderWriterImpl(Entity entity, EntityManager entityManager, ILogDispatch
 
                 public event Action<BlittableBool> Field1Updated
                 {
-                    add => field1Delegates.Add(value);
-                    remove => field1Delegates.Remove(value);
+                    add
+                    {
+                        if (LogErrorIfDisposed())
```
Comment (samiwh) [inline workers/unity/Assets/Generated/Source/improbable/gdk/tests/ExhaustiveBlittableSingularReaderWriter.cs]: In future it would be really useful to order commits for this kind of thing. If there is a huge amount of stuff that is all the same then it's much easier to read if it can be isolated to a single commit, and the rest of the code in another (or several if there are multiple logical blocks)

File: workers/unity/Assets/Generated/Source/improbable/gdk/tests/nonblittabletypes/NonBlittableComponentMonoBehaviourCommandHandlers.cs [hunk 13]
```diff
@@ -18,7 +18,8 @@ public partial class NonBlittableComponent
     {
         public partial class FirstCommand
         {
-            public struct RequestResponder {
+            public struct RequestResponder
```
Comment (samiwh) [inline workers/unity/Assets/Generated/Source/improbable/gdk/tests/nonblittabletypes/NonBlittableComponentMonoBehaviourCommandHandlers.cs]: Not really part of this PR but why is this not in the `Requirable` class?

File: workers/unity/Assets/Generated/Source/improbable/gdk/tests/ExhaustiveBlittableSingularReaderWriter.cs (hunk 12 above)
Comment (JonasImprobable) [reply workers/unity/Assets/Generated/Source/improbable/gdk/tests/ExhaustiveBlittableSingularReaderWriter.cs]: Yeah sorry about that. This was one of the situations where changes that others made in the meantime had to be pulled in and modified over and over.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReadersWriters/ReaderWriterBase.cs (hunk 11 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReadersWriters/ReaderWriterBase.cs]: Stuff seems to be still running fine because the API code doesn't depend on `ReaderWriterBase` inheriting IWriter. The final generated Impl class inherits from everything since it inherits the codegened Reader/Writer interfaces which themselves ihnehrit from the general IReader/IWriter interface. I suppose this comes down to a matter of style. I could let ReaderWriterBase inherit from IReader and IWriter if we want but it's not necessary.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReadersWriters/RequirableBase.cs (hunk 6 above)
Comment (samiwh) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReadersWriters/RequirableBase.cs]: You can reduce duplicated error spam in the logger itself. Ultimately I think we need to revisit what is an isn't an error though.

File: workers/unity/Assets/Generated/Source/improbable/gdk/tests/ExhaustiveBlittableSingularReaderWriter.cs (hunk 12 above)
Comment (samiwh) [reply workers/unity/Assets/Generated/Source/improbable/gdk/tests/ExhaustiveBlittableSingularReaderWriter.cs]: If the changes are simple then rebasing is probably fairly straightforward.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs (hunk 8 above)
Comment (samiwh) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs]: I don't agree that a system should never need to reference another one, especially as there isn't a concept of system components at the moment. However I do agree that it seems like this can be factored out. It seems like that function should be owned by what ever does the linking.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectDispatcherSystem.cs [hunk 14]
```diff
@@ -175,18 +165,30 @@ public ActivationManagerNotFoundException(string message) : base(message)
...
+
+        internal void RemoveActivationManagerAndReaderWriterStore(Entity entity)
+        {
+            if (!entityToActivationManager.TryGetValue(entity, out var activationManager))
+            {
+                throw new ActivationManagerNotFoundException($"MonoBehaviourActivationManager not found for entity {entity.Index}.");
+            }
+
+            // Disable enabled SpatialOSBehaviours and dispose leftover Requirables.
+            activationManager.Dispose();
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectDispatcherSystem.cs]: Dispose this after removing the references to it.

Comment (samiwh) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectDispatcherSystem.cs]: Also if we have disposable objects we need to have a OnDeleteManager function.
This system probably needs one anyway.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReadersWriters/RequirableBase.cs (hunk 6 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReadersWriters/RequirableBase.cs]: For the sake of getting this in quickly, are you guys ok with revisiting this later?

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs (hunk 8 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs]: We currently need to store a mapping of entity to ActivationManager somewhere. Currently it's stored in the `GameObjectDispatcherSystem`. At the same time, no matter, where the mapping is stored, the `GameObjectDispatcherSystem` needs to access it when ticked. We could start storing the mapping in the `EntityGameObjectLinker` which is currently owned by the `EntityInitializationSystem` that that would mean that the `GameObjectDispatcherSystem` needs to reference the `EntityInitializationSystem` to get the linker. Either way, as far as I can tell, all data is eventually owned by either a system or by an entity so you'll need to reference other systems somehow at the end of the day.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs (hunk 10 above)
Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs]: We can do this in a polish phase, I'll make a ticket

**Comment by JonasImprobable** [comment]
@firtina-improbable Yeah agreed. That's probably not how systemstate components should be used. Happy to make this nicer as in a follow-up refactoring step.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReadersWriters/RequirableBase.cs [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+
+        private readonly ILogDispatcher logDispatcher;
+        private bool isDisposed;
+
+        public RequirableBase(ILogDispatcher logDispatcher)
+        {
+            this.logDispatcher = logDispatcher;
+        }
+
+        protected bool LogErrorIfDisposed()
```
[the remark this answers is not in the document]
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReadersWriters/RequirableBase.cs]: Good idea. Renaming.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs (hunk 9 above)
Comment (MPeti01) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs]: This looks very confusing for an API users should be able to access and use too. I don't have much context on these buffers - could we have the Linker own them maybe and have a flush method?

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs (hunk 10 above)
Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs]: UTY-1144

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs [hunk 16]
```diff
@@ -27,6 +27,8 @@ internal class RequiredFieldInjector
...
+            = "[Require] field element does not inherit RequirableBase. This is most likely a bug in the Unity GDK.";
```
Comment (MPeti01) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs]: Couldn't we just replace the Injectable interface with the RequirableBase class? Sounds like that would make more sense. Not sure if we should do it now or just put it in a log of "cleanup refactors to do" though because it's literally everywhere though

Comment (MPeti01) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs]: Either way I don't like the error message. Firstly we are "SpatialOS GDK for Unity" now, but also maybe merge with MalformedInjectable? They are both for messed up classes that shouldn't happen at all

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReadersWriters/ReaderWriterBase.cs (hunk 11 above)
Comment (MPeti01) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReadersWriters/ReaderWriterBase.cs]: No reason why it couldn't implement the IWriter interface while also having a parent class though surely? Which would make more sense if this is where the actual method is being implemented

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs (hunk 16 above)
Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs]: 1st comment can be part of UTY-1144 - please add to the acceptance criteria

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReadersWriters/ReaderWriterBase.cs (hunk 11 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/ReadersWriters/ReaderWriterBase.cs]: Let me just let `ReaderWriterBase` from IReader and IWriter then. Why the hell not? ;)

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs (hunk 16 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs]: I don't disagree with the idea of getting rid of the `IInjectable` interface but let's do this in another PR. This PR is kinda bloated already. :)

**Comment by JonasImprobable** [comment]
@MPeti01 Re: using the `entityCommandBuffer` argument in `EntityGameObjectLinker.LinkGameObjectToEntity`: @jessicafalk 's PR should make that argument obsolete.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs [hunk 17]
```diff
@@ -43,6 +45,30 @@ public EntityGameObjectLinker(World world, ILogDispatcher logDispatcher)
             spatialOSComponent.SpatialEntityId = spatialEntityId;
             spatialOSComponent.World = world;
             spatialOSComponent.LogDispatcher = logDispatcher;
+
+            var gameObjectReference = new GameObjectReference { GameObject = gameObject };
+            viewCommandBuffer.AddComponent(entity, gameObjectReference);
+
+            var gameObjectReferenceHandleComponent = new GameObjectReferenceHandle();
+            entityCommandBuffer.AddComponent(entity, gameObjectReferenceHandleComponent);
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs]: QQ: Why do we have `GameObjectReference` and `GameObjectReferenceHandle`? I missed the context on this

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/MonoBehaviourActivationManager.cs [hunk 18]
```diff
@@ -104,26 +101,49 @@ public void EnableSpatialOSBehaviours()
...
+            DisableAllSpatialOSBehavioursOfASet(behavioursToDisable);
+            behavioursToDisable.Clear();
+        }
+
+        private void DisableAllEnabledSpatialOSBehaviours()
+        {
+            DisableAllSpatialOSBehavioursOfASet(enabledBehaviours);
+        }
+
+        private void DisableAllSpatialOSBehavioursOfASet(HashSet<MonoBehaviour> behaviours)
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/MonoBehaviourActivationManager.cs]: Nit: Can probably just rename to `DisableSpatialOSBehaviours`. The argument provides the context of the set

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs (hunk 17 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs]: At this point, there is no need for the `GameObjectReference` component anymore. We just haven't managed to remove it yet but we should eventually. :)

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs [hunk 19]
```diff
@@ -68,6 +70,28 @@ private IInjectable Inject(MonoBehaviour behaviour, InjectableId injectableId, E
...
+            {
+                var fields = fieldsToComponents.Value;
+                foreach (var field in fields)
+                {
+                    var requirableToBeDisposed = field.GetValue(behaviour) as RequirableBase;
+                    if (requirableToBeDisposed == null)
+                    {
+                        logger.HandleLog(LogType.Error, new LogEvent(RequirableFieldDoesNotInheritRequirableBase)
+                            .WithField("Behaviour", behaviour)
+                            .WithField("Field", field.GetValue(behaviour)));
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs]: What does `field.GetValue(...)` print for the user?

Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs]: This should (hopefully) call the default ToString() method of an object that was injected through the [Require] attribute.

Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs]: Wouldn't it be more useful to print the name of the field declared in the source if possible? So a user would get:

```
Field marked as required... 
Behaviour: MySpatialOSBehaviour
Field: MyIllegalRequireable
```

Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs]: You are right, that's much better. Fixed. :)

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs (hunk 17 above)
Comment (MPeti01) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs]: GameObjectReferenceHandle is removed in Jess's PR

File: workers/unity/Assets/Playground/Scenes/ClientScene.unity [hunk 20]
```diff
@@ -257,7 +257,7 @@ Prefab:
     m_Modifications:
     - target: {fileID: 4555427980518638, guid: 5857884d058555d4282a1beb501099b9, type: 2}
       propertyPath: m_LocalPosition.x
-      value: 50
```
Comment (MPeti01) [inline workers/unity/Assets/Playground/Scenes/ClientScene.unity]: What is this change?

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/MonoBehaviourActivationManager.cs (hunk 18 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/MonoBehaviourActivationManager.cs]: I renamed the method to `DisableAllSpatialOSBehavioursInternal` so people don't confuse it over the other public `DisableSpatialOSBehaviours` method, intended to be called from outside.

File: workers/unity/Packages/com.improbable.gdk.tools/.DownloadCoreSdk/Program.cs [hunk 21]
```diff
@@ -11,26 +11,26 @@ internal class Program
     {
         private static void Main(string[] args)
```
Comment (MPeti01) [inline workers/unity/Packages/com.improbable.gdk.tools/.DownloadCoreSdk/Program.cs]: What's up with all this tooling stuff showing up as changes?

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs (hunk 17 above)
Comment (jessicafalk) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs]: it will be great :D

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs (hunk 8 above)
Comment (jessicafalk) [reply workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs]: This line will be removed in my PR, so it's fine to merge it in for now

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs [hunk 22]
```diff
@@ -27,6 +27,8 @@ public class RequiredFieldInjector
             = "[Require] attribute found on member that is not Injectable. This member will be ignored.";
         private const string MalformedInjectable
             = "Injectable found without required attributes, this is invalid.";
+        private const string RequirableFieldDoesNotInheritRequirableBase
+            = "[Require] field element does not inherit RequirableBase. This is most likely a bug in the SpatialOS GDK.";
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/Injection/RequiredFieldInjector.cs]: I don't think user should be exposed to RequirableBase. It doesn't really mean anything to them.
